### PR TITLE
Merge shared parts from polysat branch

### PR DESCRIPTION
**Pull request by JakobR**

### CODE REVIEW

File: src/util/debug.h [hunk 1]
```diff
@@ -96,13 +96,15 @@ bool is_debug_enabled(const char * tag);
 
 #define VERIFY(_x_) if (!(_x_)) {                                                       \
         notify_assertion_violation(__FILE__, __LINE__, "Failed to verify: " #_x_ "\n"); \
+        DEBUG_CODE({ INVOKE_DEBUGGER(); });                                             \
```
Comment (NikolajBjorner) [inline src/util/debug.h:99]: dont port

File: src/util/debug.h [hunk 2]
```diff
@@ -96,13 +96,15 @@ bool is_debug_enabled(const char * tag);
...
         exit(ERR_UNREACHABLE);                                                          \
     }
 
 #define VERIFY_EQ(LHS, RHS)                                                                         \
     if (!((LHS) == (RHS))) {                                                                        \
         notify_assertion_violation(__FILE__, __LINE__, "Failed to verify: " #LHS " == " #RHS "\n"); \
         std::cerr << "LHS value: " << (LHS) << "\nRHS value: " << (RHS) << "\n";                    \
+        DEBUG_CODE({ INVOKE_DEBUGGER(); });                                                         \
```
Comment (NikolajBjorner) [inline src/util/debug.h:107]: dont port

**Comment by NikolajBjorner** [comment]
Better to split into pieces.
- euf_egraph.cpp changes should not be ported
- euf_egraph.h changes ok to port
- I pushed update to union_find separately
- separate PRs for tptr, vector

**Comment by JakobR** [comment]
Ok! I updated this PR and created separate ones for tptr and vector.

**Comment by nunoplopes** [comment]
LGTM!

**Pull request merged**